### ISSUE TITLE
TOC principles - ideas and suggestions

### ISSUE THREAD

**Issue by blackdwarf** [dotnet/core-docs#152]
Overview
--------

We all agree that our current TOC is not very good and needs improvement. It is also suffering from lack of direction and vision on what it needs to be and what it needs to outline going forward. This short writeup aims to remedy that, and provide suggestions for both the principles, as well as a sample of how the TOC could look like when those principles are applied. 

This is also influenced and address some thoughts from #52 . 

Goals
-----

The overall goal of our docset should be simple: to educate new and existing users on what .NET Core is and how to create applications with it. Period. 

Adding to that, our next goal should be to provide a map for new users to navigate properly through the .NET ecosystem, which, given time, will slowly migrate over to .NET Core. We now have a unique opportunity to try and set things properly vis-a-vis the richness of the overall platform and associated products. This means that we want to provide guidance to other pieces of the .NET Ecosystem that Microsoft ships when it comes to documentation. 

Principles 
----------

There are several principles I think we should adopt in documentation going forward:

* Quality over quantity
* Meticulously curated 
* Ecosystem and platform image
* Scenario organization first, component second
* Guided learning curve

Let's dig into every one of those.

**Quality over quantity** means that we don't focus on the amount of content but whether it is good or not and whether it supports the above goals. There is no need to stuff our docset with everything and the kitchen sink, it is much better to focus on those things that people have the need for in 80% of their scenarios. Also, we do have a lot of content out there, so focusing on quality first doesn't have to mean small amount of documents; for those topics that fall into 20%, we would actually direct people to existing content around the web (MSDN, CodeProject, wherever we feel the topic was best explained). 

**Meticulously curated** means that we will always be on point to curate content and that we will not accept into the core platform documentation non-curated content. All contributions will have to follow a strict set of guidelines, both in writing and topic selection; this will apply to both community and internal contribution. 

All of the documents will form an **ecosystem and platform image** insofar as they will allow consumers of the docset to actually envision how the platform is tied together and which pieces are part of the core platform and which are part of the wider ecosystem. In this way, we guarantee that people have a clear picture on the ecosystem and what is to be used when and how. 

The docset will have **scenario organization** and each specific piece of it will have to fall into one to more scenarios that were identified or new ones that will be identified. This will solve several problems that I can see with the current documentation. For instance, when you go to https://msdn.microsoft.com/en-us/library/w0x726c2(v=vs.110).aspx today, you will see a mish-mash of topics. Some will be for various versions of the .NET Framework, others will be for Entity Framework for instance. There is no mention of the actual scenario each of the component enables. So, if I'm a new user, I have no idea that "Entity Framework" is something that will actually help me with data access. If we group topics by scenarios, for instance "Working with data", we could solve this problem very easily. Once inside a given scenario, it is OK to organize by component for clarity. 

Overall organization of the documents should follow a **guided learning curve**, where first documents in a topic are aimed for beginners and last ones are aimed at advanced users. This way, we avoid having to explicitly tag documents with their "expertise level", but we also allow people a gradual introduction into any given topic. 

How would this look?
--------------------

OK, all of the above was fine and nice, but how do we get a decent TOC out of it? When we apply the principles to the TOC we have before us, we first see the following change: the current docset will raise and will have to include other components as well. If we apply the principle of scenario organization, we can see that additional libraries, such as `System.Data` or Entity Framework or WCF will be grouped by scenarios. For .NET Core, the platform and the new tooling we are building, the main scenario is building console applications. After all, these types of applications are the bedrock of everything, pretty much. Of course, I'm not talking just about the shell utilities, I'm also talking about microservices, daemons (services) etc. 

So, .NET Core would be talking about the most fundamental application scenario. Within it, we would actually have components (such as StdLib and other things that are useful). Other libraries would be part of their own scenarios, for instance "working with data", "working with formats", "networking" etc. Of course, one component may fall into many scenarios depending on the need (XML can be a format but also a data storage scenario) and that is perfectly ok. 

For those libraries that already have documentation, we link to them. For those that don't, they can live within our little universe. Those that want to move can move as well, provided that they formulate their documentation according to the principles above. 

There are those concerns which are cross-cutting from a scenario perspective, of course. One of them is "getting started", another is "working with tools". Language tutorials also fall into this category, since regardless of the application, you will actually have a need to get the toolchain, learn the languages and figure out the higher-level tools (Visual Studio, VS Code etc.)

Taking all of the above, here is how the TOC could look like (titles are examples, please do not comment on those):

* Getting started
    * Install on Linux
    * Install on OS X
    * Install on Windows
* Higher-level tools
    * Visual Studio
    * VS Code
    * Vim!
* Programming languages for .NET Core
    * C# 
    * F#
* .NET Core 
    * What is .NET Core?    
        * Overview of the platform + context
    * Writing console applications
        * Your first console application
        * How to write console applications
    * Packaging applications
        * NuGet
    * Deploying applications
    * Porting applications between editions of .NET Platform
    * .NET Command Line Interface reference 
    * Standard Library reference
        * Lib1
        * Lib2
* Working with data
    * Entity Framework
    * System.Data
* Writing servers/networking
    * System.Net
    * WCF
* Creating web applications
    * ASP.NET -> links to docs.asp.net
* Parsing formats
    * JSON
    * XML
    
And so on. This way, the TOC could grow without losing the ability to be easily navigated, given that the number of scenarios is much smaller than the number of potential technical solutions that address the needs of those scenarios.

**Comment by BethMassi** [comment]
I agree our overall documentation approach should be to get new and existing users up to speed with .NET Core. This also means looking at our valuable assets we already have built up over time on MSDN, Channel9, etc. as well. We really need to prioritize our new to .NET cross-platform developers here in core-docs because of the bias MSDN has (rightly) had to our current .NET windows developers. The content for this crowd here in this repo should be focused on understanding what the benefits of core are and how to port their existing assets. For everything else, we need to take a good look at MSDN and link there appropriately for those scenarios. We need to improve docs there too, not leave them behind. 

For new to .NET we need to have a clear learning path pop from the TOC and I don't really see that here. I would strongly recommend having a "What is .NET?" in the Getting Started that positions .NET Core and .NET framework concisely. We need to answer the "Why .NET?" right off the bat. Getting Started isn't just about installing. It's about 0-30 learning.

"Higher-level tools" should be scenario focused, not an esoteric title like this IMO. "Setting Up your Development Environment" is an important scenario that should be at the top level. 

"Programming Languages for .NET Core" again isn't scenario focused. This is part of understanding the whole platform. We need to have call to actions like "Learn C#" or if you wish "C# tutorials". The programming language portfolio should be described but that is part of the understanding of the platform IMO. 

Your scenarios, Working with data, networking, Creating Web apps, parsing formats are great and we could link to them from the language tutorials as appropriate. Things that aren't scenario focused still have a place (I'm suggesting concepts under Understanding .NET below), I would just lean on scenarios right now as we start to prioritize content building. I would change your TOC based on the topics you propose like so:

- Getting started 
    - What is .NET ? (start with http://dotnet.github.io/docs/getting-started/overview.html) 
    - Installing .NET Core 
        - Install on Linux
        - Install on OS X
        - Install on Windows
    - Setting up your Development Environment
        - Development using Visual Studio 
        - Development using Visual Studio Code
        - Development using vim
        - Development using Sublime Text
    - Writing your first console app in .NET Core  
    - Writing your first ASP.NET app using .NET Core (link to asp.net docs)
- .NET Command Line Interface reference 
-  Learn C#
    - ...Lots of C# tutorials & samples..
- Understanding .NET Core 
    - .NET Primer
        - Overview of the platform + context
    - Understanding Nuget (what is it? using it. Packaging how tos below)
    - Programming Languages on the .NET Platform 
        - C# , F#, VB, CLI
        - Roslyn
    - Asynchronous Programming with .NET
    - Generics
    - Language Integrated Query (LINQ)
    - … more concepts… 
- Creating web applications 
    - ASP.NET -> links to docs.asp.net
- Packaging applications 
    - NuGet
- Deploying applications
- Porting applications between editions of .NET Platform
- Working with data 
    - Entity Framework
    - System.Data
- Writing servers/networking 
    - System.Net
    - WCF
- Parsing formats 
    - JSON
    - XML
- Standard Library Reference
    - Library 1
    - Library 2
    - Ref docs in one place….

And remember TOCs will change over time as content builds. That's a fact in my 8 years experience building content and IA for MSDN ;-)

**Comment by ghogen** [comment]
@blackdwarf the principles you outline are great and ones that if articulated well in visible guidelines will help the community rally around a common vision.

@BethMassi I like your revisions to the proposed TOC and would extend your thinking to suggest moving the ".NET Command Line Interface reference".  What if we introduce the command line as part of Setting up your Environment (e.g. "Setting up your command line environment").  Then put the full CLI reference somewhere that doesn't interfere with the learning flow (Install, Set up environments, learn C#, orient around .NET Core, then on to specific application types...)

**Comment by BethMassi** [comment]
@ghogen Great idea. I actually thought of that after I hit the comment button and stated at the TOC a while. :smile:

**Comment by cartermp** [comment]
I like these principles a lot.  I'm reminded of [Clojure for the Brave and True](http://www.braveclojure.com/getting-started/), and although it doesn't have the same aim, it does such an excellent job getting new people started with Clojure that I feel like it has something to learn from.

I agree a bit more with @BethMassi  on what "Getting Started" encompasses.  Taking a look at what CftBaT did:

1. Describe what it is and smack you with a strong value proposition about why you need to use it.
2. Get you the artifacts you need to start writing some code.
3. Show you how to build your first app (using the most basic of tools).
4. Introduce some more tools that you may want to consider using.

So I would shuffle around @BethMassi 's "Getting Started" like this (ignore naming):

- Getting Started
   - What is .NET and why should I use it?
   - Installing (this could be one doc)
      - On Linux
      - On OS X
      - On Windows
   - How to write a console app using the most basic tooling
   - Development Environment
      - Using vs
      - Using vscode
      - Using non-microsoft editors
   - Are you a Web Developer?  Check out ASP.NET!  (link to it)
  
At this point I don't really have a strong opinion about the rest of the TOC.  I like @BethMassi 's approach for "Understanding .NET Core" acting as a set of things you should learn how to do *pronto* (and also sell some of the sexy features, like LINQ and async).  However I also think that @blackdwarf 's idea about getting some of the **really** important info out there as early as possible (like how to port) is a good approach too.

**Comment by BethMassi** [comment]
@cartermp I agree wholeheartedly with your last paragraph and @blackdwarf 's scenarios. My comments are mostly aimed at making the newbie successful quickly by proposing a natural learning plan in the TOC (not that it's the only thing there, but it should pop). Your porting docs are essential to our current devs and should also pop for that audience. I'd like to know what the MSDN peeps think about that content. How do we get the porting docs linked into MSDN? For C# tutorials, we're going to update MSDN docs in place.

**Comment by stevehoag** [comment]
@BethMassi Do you think that the porting docs belong on MSDN rather than .NET Core? Or are you thinking in both places? Probably makes more sense on MSDN and linked to from Core, since newbies aren't likely to have apps to port.

**Comment by BethMassi** [comment]
I'd lean the reverse actually. I expect .NET devs new to core to come here for docs relevant to them as well. I was thinking of flowing MSDN eyeballs to core-docs for these scenarios. In addition to audience, another major reason for that is because it's easier to update core-docs. For the C# tutorials, we're already capturing a ton of MSDN traffic there so back-porting our modern tutorials makes sense in that case.

**Comment by stevehoag** [comment]
We can easily do that - maybe just a single topic on .NET Core with links to the appropriate sections of the core docs.

**Comment by BethMassi** [comment]
I think that's a good start.

**Comment by mairaw** [comment]
We have a topic for .NET Core on MSDN but it hasn't been updated recently.
https://msdn.microsoft.com/en-US/library/dn878908(VS.110).aspx

**Comment by stevehoag** [comment]
Who owns that topic?

**Comment by ghogen** [comment]
@mairaw owns it.

**Comment by mairaw** [comment]
I do now. Cheryl wrote it.

**Comment by stevehoag** [comment]
Seems like the logical place to link into the Core docs. It definitely needs updating, and we can probably drop the link to Curah :-)

**Comment by mairaw** [comment]
It needs to be completely redone!

**Comment by blackdwarf** [comment]
OK, back to the topic of TOC. :)

I see that all of you are OK with the principles. Please be aware that the sample TOC above, in my proposal, was just that: a sample. It was not meant to be a suggestion/proposal on how the TOC should look like (as I've state, I believe). The idea of talking first about the principles is two-fold:

* We establish the spirit of the docset, so we are all on the same page what it is supposed to do
* We guide the future mutation of the TOC in such a way so we don't end up in a bad place a year or two down the road

@BethMassi I like some aspects of your TOC proposal, but dislike others. For instance, Standard Library is something that you get with .NET Core, and it shouldn't be really leveled with other scenarios, which are, essentially, separate. .NET Core is one implementaton of the .NET Platform and, given this, I think it is important to educate people on what they get "in the box", so to speak. StdLib is there any time you install .NET Core. Deploying is another thing that is tied to the app type, so it shouldn't be leveled with others.  I also don't really follow your comment about the languages...do you consider them to be an example of cross-cutting concerns or not? I would say they are, as I've stated in the proposal above. 

@cartermp @BethMassi agreed, the getting started guide should be expanded to include more things, it was rather...spartan in my sample. I would, however, leave out the topic of "write your first X app" and just assume we are showing people the ropes using console apps. It is simple, doesn't get in the way and isolated, as in, the only thing they need to learn how to do that is to have a computer and the toolchain. .NET Primer was left out of the TOC in the first issue by accident, we should definitely have it there.

**Comment by BethMassi** [comment]
@blackdwarf I thought the standard library stuff was API docs, totally agree we need to introduce the concept in the "understanding.." section. And write your first app is definitely part of the getting started experience IMO. My attempts above were aimed at taking the content we identified in the current TOC and organizing them better so we could get unblocked on writing some stuff ;-) Since we want to talk "principles" ;-) here's how I would generalize the TOC:

- Getting Started (0 to 30 hero)
    - Why bother?
    - Installing
    - Setting up
    - How to do something simple & cool
- Learn the language (30 to 60 hero)
    - Tutorial 1...n (link to concepts as appropriate, include samples)
- Understanding .NET Core
    - (0 to 60 concepts) 
    - (60 to 100 concepts)
- Scenario 1 (0 to 60)
- Scenario n (0 to 60)
- Scenario 1 (60 to 100)     
- Scenario n (60 to 100)
- Reference material (APIs, CLI reference, etc)

If we could at least get the "Learn C#" in the TOC sooner than later, we would be unblocked with the C# tutorials and could start adding them there.

**Comment by blackdwarf** [comment]
@BethMassi I don't feel that we're blocked on producing content, apart from other work (myself at least). We do know what we need, for instance the language materials, the porting guides etc. 

If we **are** blocked, then ok, let's just make a sample now and hit it. For instance, most of the scenarios that were mentioned will be driven by other teams (EF by their own team, ASP.NET by their own etc.) 

As far as the above generalization, I would just put the Reference materials in with the scenarios so they are all in the same spot. If they are not, I'm afraid we are stuck with the current MSDN model which I don't find to be helpful.

**Comment by BethMassi** [comment]
We aren't blocked on producing the content, we're blocked on putting it somewhere :-)
And reference material like APIs *should* be in one place IMO. It's fine to link to in in the conceptual docs but having a top TOC for reference material is important too.

**Comment by blackdwarf** [comment]
OK, so let me try to explain what I want to avoid; maybe there is a better way to go about it. 

I'm worried about the size of the .NET ecosystem and how that impacts our reference docs. If we have one single TOC for reference docs for **the entire ecosystem that we ship** we are stuck, I fear, in the same problem as with MSDN. Things become hard to navigate, there is no easy understanding what is in the platform and what is outside of the platform etc. 

Is my worry unfounded? Or is there a better way to solve it?

**Comment by ghogen** [comment]
When people search for reference, they'll use a search engine, so this is one reason to not worry too much about the TOC for API reference.  The goal of having a place where you can try to get a global view of the whole of the API set is good, too, but this could be conceived of as a separate piece of content than the API reference TOC structure.

**Comment by blackdwarf** [comment]
OK, so that's one solution (obviously) for finding stuff. But what about reasoning what is inside the platform and what is not?

**Comment by ghogen** [comment]
I liked the direction of your thinking at our last meeting, where you were proposing having some kind of separation, so you have the full API reference for .NET Core under one tree, but you separate out things like EF into separate trees.  Still within a reference bucket, but under a sub-bucket for, say, data scenarios.

API Reference
   * .NET Core API Reference
      * System Namespace
      * System.Text Namespace
      * Etc.
   * Entity Framework API Reference
       * System.Data.Entity
       * System.Data.Entity.Core

**Comment by BethMassi** [comment]
I like that alot too. That's what I was proposing

**Comment by cartermp** [comment]
Is having the EF API ref really our responsibility here?  It's not part of .NET Core; it's an ORM which runs on .NET.

**Comment by blackdwarf** [comment]
That's the whole point @cartermp. We are a platform, and we should provide guidance for the overall ecosystem. Therefore, my suggestion to group things by scenarios. Of course, EF (and others') docs will be written by their respective teams.

**Comment by BethMassi** [comment]
@blackdwarf exactly. We link to the scenarios and refdocs that make our platform successful. When can we act on a new TOC and start scaling out? :)

**Comment by blackdwarf** [comment]
When we all agree that we agree with the approach I outlined. :) For ref docs, we could actually group them the way @ghogen outlined (or did I on some meeting?). With that, I guess we know what our TOC looks like. I will actually make a proposal on this, if that's ok with everyone.

**Comment by BethMassi** [comment]
Yes. We're all right. ;) Can you get this TOC updated soon or do you want me to submit a PR?

**Comment by blackdwarf** [comment]
Wait, wait, we need to have all others agree first. :) 

I will be making the PR, no worries.

**Comment by mairaw** [comment]
I agree, if that counts :-)

**Comment by stevehoag** [comment]
I agree too.

**Comment by blackdwarf** [comment]
@mairaw of course it counts. :)

**Comment by BethMassi** [comment]
@mairaw and @stevehoag count double :)

**Comment by matthewreily** [comment]
I agree.  I can tell you are all very passionate about making this the best it can be and it's extremely motivating.

**Comment by cartermp** [comment]
I don't agree with this outline:

- .NET Core API Reference
   - System Namespace
   - System.Text Namespace
   - Etc.
- Entity Framework API Reference
   - System.Data.Entity
   - System.Data.Entity.Core

What I was getting at is that I don't think we should have the EF reference here, even if it's written by someone else.  It's simply not part of the platform, so it should live with [the EF docs](https://ef.readthedocs.org/en/latest/).

Is there a compelling reason why we *should* have the docs for a component like EF sitting in the API reference?

**Comment by BethMassi** [comment]
@cartermp I see what your saying about the API ref, but accessing data is a core scenario so we should at least link to it and have a scenario/conceptual doc represented here at minimum. There will be a lot of libraries that will need to plug in here. I think we should be flexible and link to guidance on our core Core scenarios.

**Comment by blackdwarf** [comment]
@cartermp also, EF docs may merge with ours at some point. This is still TBD. However, regardless of that, .NET Core is a platform, and as such, should provide information about its ecosystem to its users. Now, for your particular question, I think that having ref docs underneath its own components in scenarios would be the best way to go.

**Comment by cartermp** [comment]
@BethMassi @blackdwarf I'm not throwing out the possibility of linking to EF docs, or having a guide which uses EF to demonstrate a set of scenarios.  I just disagree with sticking the EF ref docs in this TOC, because ref docs are no longer about demonstrating the awesomeness of the platform.

As for EF docs merging with ours ... I'm concerned that if this happens, it will give momentum to the idea that other things which run on .NET Core can have their conceptual and reference docs live here.  And if that never happens because we push back (not trying to re-create MSDN after all), it just ends up being the .NET Core Platform docs + an ORM that runs on it.

**Comment by ghogen** [comment]
Is the EF we are discussing a special port of EF to .NET Core, and if so, how different is it from the desktop version? If there is not much difference, maybe the existing docs in MSDN should serve?  We would need to address the differences in some way. It could become a big problem if we set a precedent that we include a copy of the docs for every library that is ported.

**Comment by blackdwarf** [comment]
@cartermp @ghogen it doesn't really matter if the EF docs live here physically or not (and these are for EF7 so the version that works on .NET Core and also the version with enough changes that it necessitates new docs). The whole idea is for our docset to help people get an overview of the entire ecosystem, not to have them actually chase down stuff using a search engine. 

This is why I made this issue and the overall set of principles above. I believe that by following those we can actually have a good overview of the entire ecosystem without compromising the idea of the platform and actually showing our users where there are clear demarkation lines. 

Thoughts?

**Comment by cartermp** [comment]
@blackdwarf I think we will begin to compromise your principles if we start merging in EF documentation that isn't within the context of being a way to work with databases on .NET Core.  EF reference docs don't document the platform, they document an ORM framework that runs on the platform.

I just want to be clear (so that I don't continue rattling off about something under false assumptions): when you say EF docs may merge with our docs, you mean only the content which deals with certain scenarios on .NET Core, or do you mean the entire set of docs?

**Comment by blackdwarf** [comment]
@cartermp that is still TBD, I need to chat with them. However, that is, again, besides the point. So, if I understood your point above correctly, you would be for having the reference docs for EF under their own section inside bounds of a scenario (i.e. "Working with data")?

**Comment by BethMassi** [comment]
@cartermp I see what you are saying but I'm learning on @blackdwarf 's recommendation on having at least the "core" library APIs to building successful apps represented somehow in the TOC. Whether that is a link to elsewhere or in this repo, it's good for discoverability of APIs you can use. 

I would not recommend having API docs physically under scenarios (if that's what you're proposing). There is a many-to-many relationship between scenarios and APIs and scenarios will span multiple APIs and libraries. It's best to just link to them and have the scenarios pop higher up. 

Can we pull the trigger on the TOC soon? At least a V1 with a place for us to start writing? The TOC is like a project plan for the community and it'd be great if we could get something represented here soon. Fully expect this to evolve over time as content ideas come up.

**Comment by blackdwarf** [comment]
@BethMassi the .NET Core core docs (pun intended) are the minimal app you can write. The minimal app you can write is a console app that writes to STDOUT. Using EF is not, I would say, a minimal app requirement, and neither is working with relational data. However, back to my point: we are a platform and we have an ecosystem, and we need to make sure our users understand and can navigate that ecosystem. 

Also, I don't understand your point about having many-to-many relationships...even if so, wouldn't that mean that you just link?

I would not rush the TOC discussion. Or, we can do it like this:

* On this issue we continue the overall, strategic direction discussions
* On another, we iron out at least what we need in the very basic core docs and get the community started on those; we don't have to have a TOC ready for those, we can make it as an issue that they can reference in PRs

Thoughts?

**Comment by BethMassi** [comment]
Yes, that's EXACTLY my point on the many-to-many. I'm supporting you, dude. Jees! If you try and put API docs under all the scenarios *physically* it won't work because of that. I was responding to your question to @cartermp about putting EF ref docs under the scenario Working with data. I am saying "don't do that". Make sense?

Yes, let's go ahead and at least take a stab at something concrete, pretty please? That way the community is going to have a clue about our direction.

**Comment by cartermp** [comment]
@blackdwarf I would **not** be for the EF reference docs being under their section inside the bounds of a scenario.  I don't think the EF reference docs should be *anywhere* in this doc set

I think the *only* documentation that involves EF should be a conceptual doc (or docs) which describe how to work with data, using EF as a framework you could use to do that.  This doc (or docs) would link to EF docs for more advanced scenarios, the EF reference, etc.

@blackdwarf and @BethMassi w.r.t a TOC:

I think it's a good idea to pull the trigger on a v1 TOC at this point.  The most important stuff (getting started, how to write your first app, how to port, learn C#, tutorials, etc) seems like it's at a stage where we all agree.  We can iterate as we keep hashing things out in this thread and in meetings.

**Comment by BethMassi** [comment]
@cartermp hallelujah

**Comment by blackdwarf** [comment]
@cartermp well, why? What does it matter if they are there under "Working with data"? If you take a look at the proposal above, the .NET Core docs cover the "core platform" and everything else falls under the ecosystem. 

Or, let's forget about EF for a sec. Let's take, for instance, System.Xml. What if System.Xml is not part of the stdlib? Where would its docs go then? Would we not put in here because you need to reference a NuGet package to get it? 

@BethMassi again, I'm completely against pulling a trigger or anything at this point, because I don't think we know our direction yet, apart from the very basic things, which we, btw, already knew (like getting started, need for porting etc.) So, again, I don't feel we are blocked because the community can help, should they so wish, with the things we already know. Also, we do have a **ton** of issues @stevehoag created on this repo.

**Comment by BethMassi** [comment]
@blackdwarf I think you're missing the point here. People coming in reading the docs aren't necessarily looking for issues to pick up. If they see a black hole in the docs, and they say like "oh ya, I can write that" it's a LOT less friction. We need to keep up *achievable* roadblocks if we want the community to help, not difficult to navigate roadblocks.

**Comment by blackdwarf** [comment]
@BethMassi I guess I am missing the point. I was under the impression that you have people who are willing to help and are looking at what they could help with. You are saying that we will get more people if they see "black holes" in the TOC. Correct?

OK, I will make a PR to at least sort out the .NET Core portion of above, and will not put in any scenarios yet since we are not clear on how to structure those. Sounds ok?

**Comment by cartermp** [comment]
@blackdwarf I'm not sure we're communicating our points effectively.  I'm saying that I do not want any EF documentation that isn't strictly within the context of a scenario (working with data) on the platform.That means there shouldn't be any trace of reference documentation for EF in this doc set.  Not in their own separate section, not under any scenario section, not anywhere in this doc set.  The fact that EF is a part of the ecosystem doesn't mean that its reference documentation should live here as well.

In other words, the **only** docs which even mention EF should fall under the scenario of working with data, and any case where someone might want more advanced info/guidance on *how to use EF* should be linked to from a documentation site for EF.  This *other* site would also contain the EF reference.

As for `System.Xml` (or any library which isn't a part of the standard library), I think it's fine to keep the reference docs in this doc set.  To me it's a matter of size and complexity when considering what gets to have its reference docs be a part of this doc set.  

`System.Xml` is a single library that allows you to manipulate XML in a few different ways.

EF is a whole framework with many libraries, many uses, and many different scenarios.  It's scope is *far* larger than introducing someone to how you can work with a database in .NET.  Including the EF reference docs starts to make **this** doc set look a lot more like ".NET Core + EF", which is something I would want to avoid.  In other words, we're no longer documenting the platform - we're documenting the platform and Entity Framework.  I think that would be a bad move.

**Comment by blackdwarf** [comment]
@cartermp I struggle to understand how do you think that EF ref docs are not part of the "working with data" given scenario? What is so special about them? 

I guess your post above begs the question: where do we draw the line? And once we draw the line, what happens to those libraries/classes/things that are on the other side of said line? And what does that do to our overall experience of the docs?

**Comment by BethMassi** [comment]
@blackdwarf Thanks. If we can get a V1 TOC together then we can start to solicit real feedback and help from the broader community. It's much easier for people to see where we're headed and where we need help if we have a TOC. It's like a project plan. It will change based on feedback for sure. 

@cartermp I see what you are saying. At first brush, I think "how can we NOT have working with data as a scenario here? That's all I do with .NET!". But thinking about it harder, it's not really a question of not having the scenarios represented, it's about where they actually live. I could see us having a "Working with Data" topic that points to the EF documentation. Having a really close line could also help us justify not carrying every single ecosystem library, product, etc. here. Instead we could just have links to them. We should probably be focusing on CoreFX scenarios as priority 1 anyhow.

**Comment by blackdwarf** [comment]
@BethMassi I guess you and I have a different idea of what "v1" means. :) 

For all else, what I was trying to get across is that besides the purity we need to think about our users and what are their needs and how do our choices impact their experience. That is the higher-order bit.

**Comment by ghogen** [comment]
Regarding the EF 7 docs: is EF 7 is simply a newer version of the existing Framework, or is it specifically a port of EF to .NET Core?  Currently MSDN has the EF 5.0 and EF 6.0 docs [here](https://msdn.microsoft.com/en-us/library/gg696172(v=vs.103).aspx).
Also there are PMs and a doc team who own those docs, so presumably the choice of where to publish the EF reference would be up to them, and why would they not choose to publish them alongside the previous two versions, where all their existing customers would expect them, if it is simply a newer version?  If we want to proceed further with this EF discussion, we need to get the PMs and doc team for EF involved.

**Comment by blackdwarf** [comment]
@ghogen EF7 is a rewrite. A lot of things changed and yes, we will talk with their PM team this. But again, you guys are missing my point; regardless whether it is EF, WCF or something else we build in the future, we need to figure out how our platform docs give users a good experience. If it makes it easier, replace EF wth XY and think of XY as something that works on top of .NET Core that MS ships.

**Comment by BethMassi** [comment]
I'm not missing your point at all. I understand what you are saying. I don't think anyone here wants to create a crappy experience for our customers LOL. 

The last three words of your response is interesting though. When I think of "ecosystem" I do not think solely of "what MS ships". If we want that to be a "the line" here, I think that's fine if articulated clearly. But my initial impression reading your principles includes the broader ecosystem. Regardless, if we can represent the docs to these satellite libraries clearly by linking to them, that could theoretically be a good compromise. I think we need to evaluate on a case-by-case as we identify the mainline scenarios.

**Comment by BethMassi** [comment]
Or did I misread and you are saying the think MS ships is .NET Core ?

**Comment by ghogen** [comment]
@blackdwarf OK, getting away from the EF specifics.

My viewpoint is that some ported libraries will make sense to include in the .NET Core API Reference, some might not.  Whether it is included or just linked to would, I would suggest, depend on how closely the library is tied to .NET Core, whether the users are mostly .NET Core users, or mainly Windows programmers with a small number of .NET Core users for a .NET Core port of a library.

Working with XY
* Getting started with XY on the .NET Core Platform (topic contains a link to the API Ref TOC)

.NET Core API Reference
* System Namespace
* System.Text Namespace
* Etc.

XY API Reference
* XY.Z Namespace
* XY.ABC Namespace

@BethMassi I think you and I are on the same page -- many satellite libraries would be served best by links.  Some that are very .NET Core focused would or could be hosted in our TOC, and it would be a case by case choice.

**Comment by stevehoag** [comment]
+1 on this viewpoint. For example, C# is a part of many .NET Core scenarios, but I wouldn't expect to see the entire C# language reference here. For a smaller library that is primarily consumed by Core, it might make sense to pull in the docs for convenience. Case by case decisions.


Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Comment by cartermp** [comment]
@blackdwarf I think the ref docs for a component like EF *are* special, because they're no longer about the .NET Core platform anymore.  These are different because they're explicitly *not* conceptual docs, they're reference docs!  Reference docs are going to cover APIs specific to EF that are so specialized that they really don't have anything to do with the overall concept of "Working with Data".

I want to make my point as clear as possible.

**Content where we *could* bring in EF:**

* Tutorial for how to hook up an app to a SQL Database of your choice (SQL Server, Oracle, etc)
   * Maybe something about code-first or model-first if we **choose** to showcase EF
* Conceptual ideas about how to get started with setting up a data model
* etc

**EF content that _should not_ live in our docs:**

* EF API reference
* Advanced EF topics
* EF configuration tactics
* etc

There's a clear distinction here: one category is about the platform, where EF is one of many possible technologies to showcase.  The other is about a particular technology, which happens to run on the platform.  Reference docs are part of the second category.  They document a technology, not the platform.

**Comment by BethMassi** [comment]
@cartermp I was thinking for this specific case, that we would want to definitely represent the *ability* to work with data on the .NET Core platform by having a slot in the TOC "Working with Data" but that page would link directly to the EF 7 docs (like the top of the TOC there, wherever "there" is). The tutorials for "hooking up an app to a database" should live physically with the EF docs. 

Here's the way I'm looking at it so our readers have a good experience. We want to represent the fullest set of scenarios you can achieve with .NET Core so people know what the platform is capable of doing (I think that's what @blackdwarf means by "ecosystem"). That's a lot of things. If we, say, draw the line at physically carrying only docs related to learning .NET core platform and using the base class libs then we still should represent the main-line app models (EF, ASP.NET, XYZ) built on top of core that you can use to build *real applications*. That representation can just be a scenario entry in the TOC with links to those docs.

In reality, we're probably going to have people approach us to carry app model documentation so we'll have to evaluate on a case-by-case basis.

**Comment by cartermp** [comment]
@BethMassi I think that it's *totally* reasonable to have an entry in the TOC about a particular app model showing some guidance on how you can begin using it.  Where I ultimately draw the line is reference documentation or other more advanced topics which go beyond demonstrating the platform and its ecosystem, and linking to them from here.

I suspect you're going to end up finding a split of sorts if you were to go out and ask developers about if they think reference docs should live in a place like this or not.  Speaking personally, I always found it weird that MSDN had the reference docs for EF6 and MVC5, especially since I'd have to navigate back and forth between something like the [ASP.NET Identity guide](http://www.asp.net/identity/overview/getting-started/introduction-to-aspnet-identity) and the MSDN reference docs for the API I was using.

**Comment by blackdwarf** [comment]
Ok I think we will not come to an agreement on this here. I for one dont find anything "advanced" in ref docs *per se* so I fail to see how are those things, situated properly, against "demonstrating the platform and ecosystem".

However, since we have work to do I suggest we take what we agree on, which is the .NET Core core content and get cracking. I will make a PR for a new TOC tomorrow that covers only that. I also suggest we revisit this topic (scenarios and such) when we have the core stuff done to a point and see how to proceed.

**Comment by cartermp** [comment]
Agreed.  Sounds like a good plan.

**Comment by BethMassi** [comment]
Sounds like a plan

**Comment by stevehoag** [comment]
Yes, let's do it...

Sent from my Windows Phone
________________________________

**Comment by matthewreily** [comment]
+1

**Comment by mairaw** [comment]
+1 Something we all agree on! :-)

**Comment by blackdwarf** [comment]
Closing this as we have a decision and TOC is in place.
